File: scripts/dev_tools/codex_native_converter/validation.py
"""Validate planned conversions for the Codex-native converter.

Purpose:
    Enforce the converter's fail-closed rules before apply mode writes native
    output into a destination workspace.

Usage:
    The converter engine calls ``validate_conversion_plan`` after mapping and
    rewrite application and before report emission decides whether apply mode may
    proceed.

Flow:
    Validation checks required inputs first, then inspects mapping records for
    unsupported or unresolved states, then inspects generated output for
    duplicate targets and lingering runtime references.

Invariants / Constraints:
    Blocking failures are reported explicitly with stable codes so review and
    apply mode share one auditable validation contract.

Side Effects:
    None.
"""

from __future__ import annotations

from collections import Counter

from scripts.dev_tools.codex_native_converter.models import (
    ConversionClass,
    MappingRecord,
    PlannedEmission,
    RunOptions,
    TargetRole,
    ValidationFinding,
)
from scripts.dev_tools.codex_native_converter.rewrites import (
    detect_unresolved_runtime_reference,
)

_NOTE_FLAG_TO_FINDING: tuple[tuple[str, str, str], ...] = (
    (
        "requires-native-hard-gate",
        "unresolved-hard-gate-mapping",
        "Source artifact requires a native hard-gate mapping that is not yet resolved.",
    ),
    (
        "requires-handoff-review",
        "unresolved-handoff-mapping",
        "Source artifact requires handoff or delegation behavior that is not "
        "yet resolved.",
    ),
    (
        "requires-mcp-rewrite",
        "unresolved-mcp-rewrite",
        "Source artifact requires a semantic MCP rewrite that is not yet resolved.",
    ),
    (
        "malformed-source-artifact",
        "malformed-source-artifact",
        "Source artifact metadata or content is malformed for v1 conversion.",
    ),
)


def _build_finding(
    *,
    code: str,
    blocking: bool,
    source_path: str | None,
    target_path: str | None,
    message: str,
    recommended_action: str,
) -> ValidationFinding:
    """Build a validation finding with consistent severity values.

    Purpose:
        Keep finding construction consistent across all validation branches.

    Args:
        code (str): Stable machine-readable validation code.
        blocking (bool): Whether the finding must stop apply mode.
        source_path (str | None): Source-relative path associated with the
            finding.
        target_path (str | None): Target-relative path associated with the
            finding.
        message (str): Human-readable explanation.
        recommended_action (str): Human-readable remediation guidance.

    Returns:
        ValidationFinding: The normalized validation finding.

    Raises:
        None.

    Side Effects:
        None.
    """

    return ValidationFinding(
        code=code,
        severity="error" if blocking else "warning",
        blocking=blocking,
        source_path=source_path,
        target_path=target_path,
        message=message,
        recommended_action=recommended_action,
    )


def _validate_required_inputs(run_options: RunOptions) -> list[ValidationFinding]:
    """Validate run options that gate review and apply execution.

    Purpose:
        Fail early when the caller omitted required inputs.

    Args:
        run_options (RunOptions): Requested converter run options.

    Returns:
        list[ValidationFinding]: Findings for missing required inputs.

    Raises:
        None.

    Side Effects:
        None.
    """

    findings: list[ValidationFinding] = []
    if run_options.mode == "apply" and run_options.destination_root is None:
        findings.append(
            _build_finding(
                code="missing-required-input",
                blocking=True,
                source_path=None,
                target_path=None,
                message="Apply mode requires an explicit destination root.",
                recommended_action=(
                    "Provide destination_root before running apply mode."
                ),
            )
        )
    return findings


def _validate_mapping_records(
    mapping_records: tuple[MappingRecord, ...],
) -> list[ValidationFinding]:
    """Validate classified and mapped records for unsupported or flagged states.

    Purpose:
        Convert unsupported or unresolved record states into explicit findings.

    Args:
        mapping_records (tuple[MappingRecord, ...]): Planned mappings for the
            current run.

    Returns:
        list[ValidationFinding]: Findings derived from mapping records.

    Raises:
        None.

    Side Effects:
        None.
    """

    findings: list[ValidationFinding] = []

    for mapping_record in mapping_records:
        if (
            mapping_record.conversion_class is ConversionClass.UNSUPPORTED
            or mapping_record.target_role is TargetRole.UNSUPPORTED
        ) and mapping_record.is_required:
            findings.append(
                _build_finding(
                    code="unsupported-ecosystem",
                    blocking=True,
                    source_path=mapping_record.source_path,
                    target_path=mapping_record.target_path,
                    message="Required source artifact has no safe v1 mapping.",
                    recommended_action=(
                        "Review the artifact manually or add a verified "
                        "native mapping before apply mode."
                    ),
                )
            )

        for note in mapping_record.notes:
            for note_flag, finding_code, finding_message in _NOTE_FLAG_TO_FINDING:
                if note_flag in note:
                    findings.append(
                        _build_finding(
                            code=finding_code,
                            blocking=True,
                            source_path=mapping_record.source_path,
                            target_path=mapping_record.target_path,
                            message=finding_message,
                            recommended_action=(
                                "Resolve the flagged mapping or leave the "
                                "artifact in review-only status."
                            ),
                        )
                    )

    return findings


def _validate_duplicate_targets(
    mapping_records: tuple[MappingRecord, ...],
    planned_emissions: tuple[PlannedEmission, ...],
) -> list[ValidationFinding]:
    """Validate that planned target paths are unique.

    Purpose:
        Prevent ambiguous or silently overwritten target outputs.

    Args:
        mapping_records (tuple[MappingRecord, ...]): Planned mappings for the
            current run.
        planned_emissions (tuple[PlannedEmission, ...]): Section-level planned
            emissions that may also claim target paths.

    Returns:
        list[ValidationFinding]: Findings for duplicate target paths.

    Raises:
        None.

    Side Effects:
        None.
    """

    findings: list[ValidationFinding] = []
    mapping_records_by_target: dict[str, list[MappingRecord]] = {}
    section_emissions_by_target: dict[str, list[PlannedEmission]] = {}

    for mapping_record in mapping_records:
        if mapping_record.target_path is None:
            continue
        mapping_records_by_target.setdefault(mapping_record.target_path, []).append(
            mapping_record
        )
    for planned_emission in planned_emissions:
        if planned_emission.target_path is None:
            continue
        section_emissions_by_target.setdefault(planned_emission.target_path, []).append(
            planned_emission
        )

    target_counter = Counter(
        mapping_record.target_path
        for mapping_record in mapping_records
        if mapping_record.target_path is not None
    )
    duplicated_targets = {
        target_path for target_path, count in target_counter.items() if count > 1
    }
    duplicated_targets -= {
        target_path
        for target_path, records in mapping_records_by_target.items()
        if target_path == "AGENTS.md"
        and all(
            record.target_role is TargetRole.STANDING_GUIDANCE for record in records
        )
    }

    # Report every record that participates in a duplicate target collision so
    # the caller can trace all conflicting sources.
    for mapping_record in mapping_records:
        if mapping_record.target_path in duplicated_targets:
            findings.append(
                _build_finding(
                    code="duplicate-target-path",
                    blocking=True,
                    source_path=mapping_record.source_path,
                    target_path=mapping_record.target_path,
                    message=(
                        "Multiple source artifacts resolve to the same target " "path."
                    ),
                    recommended_action=(
                        "Refine the mapping plan so each target path has "
                        "exactly one authoritative source."
                    ),
                )
            )

    conflicting_targets = {
        target_path
        for target_path, section_emissions in section_emissions_by_target.items()
        if (
            target_path in mapping_records_by_target
            or len(
                {
                    (planned_emission.source_path, planned_emission.target_role)
                    for planned_emission in section_emissions
                }
            )
            > 1
        )
    }

    for target_path in sorted(conflicting_targets):
        for planned_emission in section_emissions_by_target[target_path]:
            findings.append(
                _build_finding(
                    code="duplicate-target-path",
                    blocking=True,
                    source_path=planned_emission.source_path,
                    target_path=target_path,
                    message=(
                        "Multiple planned emissions resolve to the same target path."
                    ),
                    recommended_action=(
                        "Refine the section-level mapping plan so each target path has "
                        "exactly one authoritative emission group."
                    ),
                )
            )
    return findings


def _validate_generated_output(
    generated_output: dict[str, str],
) -> list[ValidationFinding]:
    """Validate generated output text for lingering runtime references.

    Purpose:
        Ensure emitted native Codex outputs do not retain unresolved source
        runtime references.

    Args:
        generated_output (dict[str, str]): Generated output keyed by target path.

    Returns:
        list[ValidationFinding]: Findings for lingering runtime references.

    Raises:
        None.

    Side Effects:
        None.
    """

    findings: list[ValidationFinding] = []

    # Scan every generated target body after rewrites so apply mode fails closed
    # when native outputs still mention source-runtime surfaces.
    for target_path, rendered_text in generated_output.items():
        unresolved_references = detect_unresolved_runtime_reference(rendered_text)
        if unresolved_references:
            findings.append(
                _build_finding(
                    code="lingering-source-runtime-reference",
                    blocking=True,
                    source_path=None,
                    target_path=target_path,
                    message=(
                        "Generated output retains unresolved source-runtime "
                        "references: " + ", ".join(unresolved_references)
                    ),
                    recommended_action=(
                        "Add a verified rewrite or adjust the generated "
                        "content before apply mode."
                    ),
                )
            )

    return findings


def validate_conversion_plan(
    run_options: RunOptions,
    mapping_records: tuple[MappingRecord, ...],
    planned_emissions: tuple[PlannedEmission, ...],
    generated_output: dict[str, str],
) -> tuple[ValidationFinding, ...]:
    """Validate one planned conversion run.

    Purpose:
        Apply the converter's full fail-closed validation contract to the mapped
        artifacts and generated output for one run.

    Args:
        run_options (RunOptions): Requested converter run options.
        mapping_records (tuple[MappingRecord, ...]): Planned mappings for the
            current run.
        planned_emissions (tuple[PlannedEmission, ...]): Section-level planned
            emissions for the current run.
        generated_output (dict[str, str]): Generated output keyed by target path.

    Returns:
        tuple[ValidationFinding, ...]: Validation findings sorted by stable key.

    Raises:
        None.

    Side Effects:
        None.
    """

    findings: list[ValidationFinding] = []
    findings.extend(_validate_required_inputs(run_options))
    findings.extend(_validate_mapping_records(mapping_records))
    findings.extend(_validate_duplicate_targets(mapping_records, planned_emissions))
    findings.extend(_validate_generated_output(generated_output))

    return tuple(
        sorted(
            findings,
            key=lambda finding: (
                finding.code,
                finding.source_path or "",
                finding.target_path or "",
            ),
        )
    )

File: extensions/drm-copilot/src/lib/codex-native-converter/validation.ts
/**
 * Validate planned conversions for the Codex-native converter.
 *
 * Purpose:
 *     Enforce the converter's fail-closed rules before apply mode writes native
 *     output. Ported from `validation.py`; pure logic, no I/O.
 *
 * Invariants:
 *     Blocking failures are reported explicitly with stable codes so review and
 *     apply mode share one auditable validation contract. Findings are returned
 *     in a deterministic sorted order.
 */

import {
  ConversionClass,
  type MappingRecord,
  type PlannedEmission,
  type RunOptions,
  TargetRole,
  type ValidationFinding,
} from "./models";
import { detectUnresolvedRuntimeReference } from "./rewrites";

// Note-flag substrings mapped to the finding code and message they raise,
// mirroring `_NOTE_FLAG_TO_FINDING`.
const NOTE_FLAG_TO_FINDING: ReadonlyArray<readonly [string, string, string]> = [
  [
    "requires-native-hard-gate",
    "unresolved-hard-gate-mapping",
    "Source artifact requires a native hard-gate mapping that is not yet resolved.",
  ],
  [
    "requires-handoff-review",
    "unresolved-handoff-mapping",
    "Source artifact requires handoff or delegation behavior that is not " +
      "yet resolved.",
  ],
  [
    "requires-mcp-rewrite",
    "unresolved-mcp-rewrite",
    "Source artifact requires a semantic MCP rewrite that is not yet resolved.",
  ],
  [
    "malformed-source-artifact",
    "malformed-source-artifact",
    "Source artifact metadata or content is malformed for v1 conversion.",
  ],
];

/**
 * Build a validation finding with a severity derived from the blocking flag.
 *
 * Mirrors `_build_finding`.
 *
 * @param options Finding code/blocking/paths/message/recommendedAction.
 * @returns The normalized validation finding.
 */
function buildFinding(options: {
  readonly code: string;
  readonly blocking: boolean;
  readonly sourcePath: string | null;
  readonly targetPath: string | null;
  readonly message: string;
  readonly recommendedAction: string;
}): ValidationFinding {
  return {
    code: options.code,
    severity: options.blocking ? "error" : "warning",
    blocking: options.blocking,
    sourcePath: options.sourcePath,
    targetPath: options.targetPath,
    message: options.message,
    recommendedAction: options.recommendedAction,
  };
}

/**
 * Validate run options that gate review and apply execution.
 *
 * Mirrors `_validate_required_inputs`.
 *
 * @param runOptions Requested converter run options.
 * @returns Findings for missing required inputs.
 */
function validateRequiredInputs(runOptions: RunOptions): ValidationFinding[] {
  const findings: ValidationFinding[] = [];
  if (runOptions.mode === "apply" && runOptions.destinationRoot === null) {
    findings.push(
      buildFinding({
        code: "missing-required-input",
        blocking: true,
        sourcePath: null,
        targetPath: null,
        message: "Apply mode requires an explicit destination root.",
        recommendedAction:
          "Provide destination_root before running apply mode.",
      }),
    );
  }
  return findings;
}

/**
 * Validate classified and mapped records for unsupported or flagged states.
 *
 * Mirrors `_validate_mapping_records`.
 *
 * @param mappingRecords Planned mappings for the current run.
 * @returns Findings derived from mapping records.
 */
function validateMappingRecords(
  mappingRecords: ReadonlyArray<MappingRecord>,
): ValidationFinding[] {
  const findings: ValidationFinding[] = [];

  for (const mappingRecord of mappingRecords) {
    // A required artifact with no safe mapping is a blocking finding.
    if (
      (mappingRecord.conversionClass === ConversionClass.UNSUPPORTED ||
        mappingRecord.targetRole === TargetRole.UNSUPPORTED) &&
      mappingRecord.isRequired
    ) {
      findings.push(
        buildFinding({
          code: "unsupported-ecosystem",
          blocking: true,
          sourcePath: mappingRecord.sourcePath,
          targetPath: mappingRecord.targetPath,
          message: "Required source artifact has no safe v1 mapping.",
          recommendedAction:
            "Review the artifact manually or add a verified " +
            "native mapping before apply mode.",
        }),
      );
    }

    // Each note may carry a flag that raises a specific unresolved finding.
    for (const note of mappingRecord.notes) {
      for (const [
        noteFlag,
        findingCode,
        findingMessage,
      ] of NOTE_FLAG_TO_FINDING) {
        if (note.includes(noteFlag)) {
          findings.push(
            buildFinding({
              code: findingCode,
              blocking: true,
              sourcePath: mappingRecord.sourcePath,
              targetPath: mappingRecord.targetPath,
              message: findingMessage,
              recommendedAction:
                "Resolve the flagged mapping or leave the " +
                "artifact in review-only status.",
            }),
          );
        }
      }
    }
  }

  return findings;
}

/**
 * Validate that planned target paths are unique.
 *
 * Mirrors `_validate_duplicate_targets`, including the `AGENTS.md`
 * standing-guidance exemption and the section-emission conflict detection.
 *
 * @param mappingRecords Planned mappings for the current run.
 * @param plannedEmissions Section-level planned emissions.
 * @returns Findings for duplicate target paths.
 */
function validateDuplicateTargets(
  mappingRecords: ReadonlyArray<MappingRecord>,
  plannedEmissions: ReadonlyArray<PlannedEmission>,
): ValidationFinding[] {
  const findings: ValidationFinding[] = [];
  const mappingRecordsByTarget = new Map<string, MappingRecord[]>();
  const sectionEmissionsByTarget = new Map<string, PlannedEmission[]>();

  // Group mapping records and planned emissions by their target path.
  for (const mappingRecord of mappingRecords) {
    if (mappingRecord.targetPath === null) {
      continue;
    }
    const group = mappingRecordsByTarget.get(mappingRecord.targetPath) ?? [];
    group.push(mappingRecord);
    mappingRecordsByTarget.set(mappingRecord.targetPath, group);
  }
  for (const plannedEmission of plannedEmissions) {
    if (plannedEmission.targetPath === null) {
      continue;
    }
    const group =
      sectionEmissionsByTarget.get(plannedEmission.targetPath) ?? [];
    group.push(plannedEmission);
    sectionEmissionsByTarget.set(plannedEmission.targetPath, group);
  }

  // Count mapping-record targets to find duplicates.
  const targetCounts = new Map<string, number>();
  for (const mappingRecord of mappingRecords) {
    if (mappingRecord.targetPath === null) {
      continue;
    }
    targetCounts.set(
      mappingRecord.targetPath,
      (targetCounts.get(mappingRecord.targetPath) ?? 0) + 1,
    );
  }
  const duplicatedTargets = new Set<string>();
  for (const [targetPath, count] of targetCounts) {
    if (count > 1) {
      duplicatedTargets.add(targetPath);
    }
  }
  // AGENTS.md is exempt when every record targeting it is standing guidance.
  for (const [targetPath, records] of mappingRecordsByTarget) {
    if (
      targetPath === "AGENTS.md" &&
      records.every(
        (record) => record.targetRole === TargetRole.STANDING_GUIDANCE,
      )
    ) {
      duplicatedTargets.delete(targetPath);
    }
  }

  // Report every record participating in a duplicate target collision.
  for (const mappingRecord of mappingRecords) {
    if (
      mappingRecord.targetPath !== null &&
      duplicatedTargets.has(mappingRecord.targetPath)
    ) {
      findings.push(
        buildFinding({
          code: "duplicate-target-path",
          blocking: true,
          sourcePath: mappingRecord.sourcePath,
          targetPath: mappingRecord.targetPath,
          message:
            "Multiple source artifacts resolve to the same target " + "path.",
          recommendedAction:
            "Refine the mapping plan so each target path has " +
            "exactly one authoritative source.",
        }),
      );
    }
  }

  // A section-emission target conflicts when it also has a mapping record, or
  // when its emissions span more than one (source_path, target_role) pair.
  const conflictingTargets = new Set<string>();
  for (const [targetPath, sectionEmissions] of sectionEmissionsByTarget) {
    const distinctPairs = new Set(
      sectionEmissions.map(
        (emission) => `${emission.sourcePath}\u0000${emission.targetRole}`,
      ),
    );
    if (mappingRecordsByTarget.has(targetPath) || distinctPairs.size > 1) {
      conflictingTargets.add(targetPath);
    }
  }

  // Report conflicting section emissions in sorted target-path order.
  for (const targetPath of [...conflictingTargets].sort((left, right) =>
    left < right ? -1 : left > right ? 1 : 0,
  )) {
    for (const plannedEmission of sectionEmissionsByTarget.get(targetPath) ??
      []) {
      findings.push(
        buildFinding({
          code: "duplicate-target-path",
          blocking: true,
          sourcePath: plannedEmission.sourcePath,
          targetPath,
          message:
            "Multiple planned emissions resolve to the same target path.",
          recommendedAction:
            "Refine the section-level mapping plan so each target path has " +
            "exactly one authoritative emission group.",
        }),
      );
    }
  }

  return findings;
}

/**
 * Validate generated output text for lingering runtime references.
 *
 * Mirrors `_validate_generated_output`.
 *
 * @param generatedOutput Generated output keyed by target path.
 * @returns Findings for lingering runtime references.
 */
function validateGeneratedOutput(
  generatedOutput: Readonly<Record<string, string>>,
): ValidationFinding[] {
  const findings: ValidationFinding[] = [];

  // Scan every generated target body after rewrites so apply mode fails closed
  // when native outputs still mention source-runtime surfaces.
  for (const targetPath of Object.keys(generatedOutput)) {
    const renderedText = generatedOutput[targetPath] ?? "";
    const unresolvedReferences = detectUnresolvedRuntimeReference(renderedText);
    if (unresolvedReferences.length > 0) {
      findings.push(
        buildFinding({
          code: "lingering-source-runtime-reference",
          blocking: true,
          sourcePath: null,
          targetPath,
          message:
            "Generated output retains unresolved source-runtime " +
            "references: " +
            unresolvedReferences.join(", "),
          recommendedAction:
            "Add a verified rewrite or adjust the generated " +
            "content before apply mode.",
        }),
      );
    }
  }

  return findings;
}

/**
 * Validate one planned conversion run.
 *
 * Mirrors `validate_conversion_plan`: applies the required-input, mapping,
 * duplicate-target, and generated-output checks, then sorts findings by
 * (code, sourcePath, targetPath) with null paths treated as the empty string.
 *
 * @param runOptions Requested converter run options.
 * @param mappingRecords Planned mappings for the current run.
 * @param plannedEmissions Section-level planned emissions for the current run.
 * @param generatedOutput Generated output keyed by target path.
 * @returns Validation findings sorted by stable key.
 */
export function validateConversionPlan(
  runOptions: RunOptions,
  mappingRecords: ReadonlyArray<MappingRecord>,
  plannedEmissions: ReadonlyArray<PlannedEmission>,
  generatedOutput: Readonly<Record<string, string>>,
): ReadonlyArray<ValidationFinding> {
  const findings: ValidationFinding[] = [
    ...validateRequiredInputs(runOptions),
    ...validateMappingRecords(mappingRecords),
    ...validateDuplicateTargets(mappingRecords, plannedEmissions),
    ...validateGeneratedOutput(generatedOutput),
  ];

  // Sort by (code, sourcePath, targetPath); null paths sort as empty strings.
  const compare = (left: string, right: string): number =>
    left < right ? -1 : left > right ? 1 : 0;
  return [...findings].sort((left, right) => {
    const byCode = compare(left.code, right.code);
    if (byCode !== 0) {
      return byCode;
    }
    const bySource = compare(left.sourcePath ?? "", right.sourcePath ?? "");
    if (bySource !== 0) {
      return bySource;
    }
    return compare(left.targetPath ?? "", right.targetPath ?? "");
  });
}
